rep-2004.rst

```diff
@@ -0,0 +1,832 @@
+However, a package which provides pose estimation (like ``robot_pose_ekf``\ ) is a generic solution for something that most people need, but is often replaced with a domain specific solution in production, and therefore it should probably not start out as 'Level 1'.
+Such a package may upgrade to 'Level 1' at a later date, if it proves to be a solution that people want to use in their products.
+
+Tools, like ``rostopic``\ , generally do not fall into this category either, but are not categorically excluded.
+For example, it may be the case that the tool which launches and verifies a ROS graph (``ros2launch``\ ) may need to be considered 'Level 1' for use in production systems.
+
+Package Requirements
+~~~~~~~~~~~~~~~~~~~~
+
+Requirements to be considered a 'Level 1' package:
+
+.. _Version Policy:
+
+1. **Version Policy:**
+
+   .. _1.i:
+
+   i. Must have a version policy (e.g. ``semver``)
+
+   .. _1.ii:
+
+   ii. Must be at a stable version (e.g. for ``semver`` that means version >= 1.0.0)
+
+   .. _1.iii:
+
+   iii. Must have a strictly declared public API
+
+   .. _1.iv:
+
+   iv. Must have a policy for API stability
+
+   .. _1.v:
+
+   v. Must have a policy for ABI stability
+
+   .. _1.vi:
+
+   vi. Must have a policy that keeps API and ABI stability within a released ROS distribution
+
+.. _Change Control Process:
+
+2. **Change Control Process:**
+
+   .. _2.i:
+
+   i. Must have all code changes occur through a change request (e.g. pull request, merge request, etc.)
+
+   .. _2.ii:
+
+   ii. Must have confirmation of contributor origin (e.g. `DCO  <https://developercertificate.org/>`_, CLA, etc.)
+
+   .. _2.iii:
+
+   iii. Must have peer review policy for all change requests (e.g. require one or more reviewers)
+
+   .. _2.iv:
+
+   iv. Must have Continuous Integration (CI) policy for all change requests
+
+   .. _2.v:
+
+   v. Must have documentation policy for all change requests
+
+.. _Documentation:
+
+3. **Documentation:**
+
+   .. _3.i:
+
+   i. Must have documentation for each "feature" (e.g. for ``rclcpp``: create a node, publish a message, spin, etc.)
+
+   .. _3.ii:
+
+   ii. Must have documentation for each item in the public API (e.g. functions, classes, etc.)
+
+   .. _3.iii:
+
+   iii. Must have a declared license or set of licenses
+
+   .. _3.iv:
+
+   iv. Must have a copyright statement in each source file
+
+   .. _3.v:
+
+   v. Must have a "quality declaration" document, which declares the quality level and justifies how the package meets each of the requirements
+
+      .. _3.v.a:
+
+      a. *Must have a section in the repository's ``README`` which contains the "quality declaration" or links to it*
+
+      .. _3.v.b:
+
+      b. *Must register with a centralized list of 'Level N' packages, if one exists, to allow for peer review of the claim*
```

Comment:
```suggestion
      b. *Must allow peer review of this claim and, if one exists, register with a centralized list of 'Level N' packages*
```
I think the important part of this section is making clear that a QD claim can be reviewed publicly. As the document says:

> There should be one or more communal lists of 'Level 1' (and maybe 'Level 2' or 'Level 3') quality level packages.

This implies there may not be a list for all the package levels. The wording change suggestion will allow flexibility around the centralized list and making emphasis around the peer review part.

Reply:
I don't feel like this is any better or worse than the current line. These two things are connected, so listing one before the other doesn't seem to change the importance to me.

Like "I must buy food therefore I must go to the store" versus "I must go to the store, so that I can buy food", they don't indicate which is more important.

@maryaB-osr what do you think?

Reply:
Also, with the new wording it sounds like disallowing peer review was an option, which is not really the case,  you cannot stop peer review of your open source project. Instead I would say "Must seek peer review...", but that's what you are trying to do by registering with a centralized list...

Reply:
The suggestion is incorrect because it changes the target of "_if ones exists_" to the "_this claim_" immediately preceding it. We can understand it because we know that the Level N lists may or may not exist, but someone reading that for the first time could easily mistake the condition to be whether a claim/QD exists.

Reply:
@maryaB-osr I agree with both of you, my wording was not correct. I mainly wanted to raise a couple of concerns related to the "Centralized list":

1. If that list doesn't exist or if the package is not registered in that list, will that affect the quality stated by the package? (even if their authors explicitly state in their Quality Declarations they are open for peer reviews of their claims?)

2. If that list exists, but the package claims are updated faster than the list (as I suppose that list will require some sort of maintenance), will those claims be invalid until the list is updated?

My suggestion is to make emphasis in allowing peer review while making the list as something secondary in comparison:
*Must allow peer review of the quality level claims. And also should register with a centralized list of 'Level N' packages (if one exists)*

Reply:
> If that list doesn't exist or if the package is not registered in that list, will that affect the quality stated by the package? (even if their authors explicitly state in their Quality Declarations they are open for peer reviews of their claims?)

No, I don't see why it would.

> If that list exists, but the package claims are updated faster than the list (as I suppose that list will require some sort of maintenance), will those claims be invalid until the list is updated?

No, the list should not just include which packages, but which versions of the package are in one quality level or another. If the package makes improvements to their quality level without releasing a new version of their package, then they could ask for a re-evaluation on one list or another.

I think what you confusion is from is that the list is entirely supplemental to the quality declaration. The package maintainers must have a QD and must declare that they are in a certain level. A user can use that information directly and evaluate the claims in their QD themselves. I'd recommend this path for any users with serious quality or safety concerns, i.e. trust but independently verify.

Completely separately, we can (and should) have peer-reviewed lists of packages which confirm the packages are in the quality level they declare. More casual users may "trust" or "depend on" these lists rather than verifying the claims themselves. But then they are explicitly trusting the maintainers of the list instead, and in exchange they have a lot less work to do. But again, I'll emphasize, the peer review is entirely optional and the curated lists are entirely optional, and the quality level of the package and the quality declaration do not depend on these lists or peer reviews. We recommend that package maintainers seek peer review (primarily by trying to get on one of these peer reviewed lists), but we have no way of enforcing that.

With respect to:

> My suggestion is to make emphasis in allowing peer review while making the list as something secondary in comparison:
Must allow peer review of the quality level claims. And also should register with a centralized list of 'Level N' packages (if one exists)

I'll just quote myself, since it doesn't make sense to me:

> Also, with the new wording it sounds like disallowing peer review was an option, which is not really the case, you cannot stop peer review of your open source project. Instead I would say "Must seek peer review...", but that's what you are trying to do by registering with a centralized list...

Reply:
> 1. If that list doesn't exist or if the package is not registered in that list, will that affect the quality stated by the package? (even if their authors explicitly state in their Quality Declarations they are open for peer reviews of their claims?)

@Blast545 I think you're putting too much emphasis on the "peer review" part of this requirement. This is the only requirement that mentions peer review, and it's only part of a conditional: "_If_ the list exists, then the package must register with it, _in turn_ allowing for peer review". Peer review is just a side effect of registering with the list. As soon as "If the list exists" is not true, the rest of the statement is not a requirement anymore. There's no responsibility to explicitly seek review besides using these lists. The "Quality Declaration and Claim" section below makes this clear.

I would say that if the package is not registered with an existing list, then yes that would affect the quality level. Otherwise, what is 3.v.b. qualifying?

@wjwwood I think where you said:

> ...the peer review is entirely optional and the curated lists are entirely optional, and the quality level of the package and the quality declaration do not depend on these lists or peer reviews.

That should mean "entirely optional _if the list doesn't exist_"? If you mean "optional even if the list exists" then this needs to say "**Should** register with a centralized list..." and needs to be marked as a recommendation (●) and not a requirement (✓) in the table.

Reply:
I think "Should register with a centralized list..." is appropriate. I think there's value in the QD and the self declared quality level even if they're not part of a list. It also causes a catch 22, where you cannot apply to the list until you have declared your level and you cannot declare level 1 because you are not on the list yet.

I would move to make adding yourself to a list, the list, all the lists, etc... to be a recommendation and not a requirement.

What makes sense to me as a requirement is that if your package is a part of one or more lists that you reference them in your QD.

Reply:
> I would say that if the package is not registered with an existing list, then yes that would affect the quality level. Otherwise, what is 3.v.b. qualifying?

I think maybe 3.v.b is unnecessary (my fault for including it). As I said above, I think that recommending maintainers try to register with a list and generally seek peer review is good enough. If there's any require it is to list the peer reviewing and lists your package has had in your QD, but only to make it easier for consumers of the QD to connect the dots.

Reply:
Yes, it wasn't clear to me the purpose of this section.

> Completely separately, we can (and should) have peer-reviewed lists of packages which confirm the packages are in the quality level they declare. More casual users may "trust" or "depend on" these lists rather than verifying the claims themselves. But then they are explicitly trusting the maintainers of the list instead, and in exchange they have a lot less work to do.

This sounds good to me!